### PR TITLE
Fix windows service wrapper

### LINKED ISSUE #83270: Any -XLog option (from env of options file) is removed in windows service wrapper script
### Elasticsearch Version

<= 8

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Windows

### Problem Description

Any -XLog option (probably containing a coma or wildcard) like the native one `-Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m` is systematicaly removed from `OTHER_JAVA_OPTS` here https://github.com/elastic/elasticsearch/blob/master/distribution/src/bin/elasticsearch-service.bat#L178

The value inside `ES_JAVA_OPTS` just before the loop is correct ! The bug is in the batch loop implementation.

### Steps to Reproduce

Just install the service on windows with `./elasticsearch-service.bat install ElasticSearch`
Then open the Procrun manager `elasticsearch-service.bat manager elasticsearch`

See in Java tab, XLog arguments are absent !

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by gillg**
Fix https://github.com/elastic/elasticsearch/issues/83270

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

Sample command to create the service before the fix :
`"D:\elasticsearch\release\bin\elasticsearch-service-x64.exe" //IS//ElasticSearch --Startup manual --StopTimeout 0 --StartClass org.elasticsearch.bootstrap.Elasticsearch --StartMethod main ++StartParams --quiet --StopClass org.elasticsearch.bootstrap.Elasticsearch --StopMethod close --Classpath "D:\elasticsearch\release\lib\*" --JvmMs 4192  --JvmMx 4192  --JvmSs 1024  --JvmOptions -Des.networkaddress.cache.ttl=60;-Des.networkaddress.cache.negative.ttl=10;-XX:+AlwaysPreTouch;-Djava.awt.headless=true;-Dfile.encoding=UTF-8;-Djna.nosys=true;-XX:-OmitStackTraceInFastThrow;-XX:+ShowCodeDetailsInExceptionMessages;-Dio.netty.noUnsafe=true;-Dio.netty.noKeySetOptimization=true;-Dio.netty.recycler.maxCapacityPerThread=0;-Dio.netty.allocator.numDirectArenas=0;-Dlog4j.shutdownHookEnabled=false;-Dlog4j2.disable.jmx=true;-Dlog4j2.formatMsgNoLookups=true;-Djava.locale.providers=SPI,COMPAT;--add-opens=java.base/java.io=ALL-UNNAMED;-XX:+UseG1GC;-XX:G1ReservePercent=25;-XX:InitiatingHeapOccupancyPercent=30;-Djava.io.tmpdir=D:\elasticsearch\\release\..\temp;-XX:+HeapDumpOnOutOfMemoryError;-XX:HeapDumpPath=data;-XX:ErrorFile=logs/hs_err_pid%p.log;-Dlog4j2.formatMsgNoLookups=true;-XX:HeapDumpPath=D:\elasticsearch\data;-XX:ErrorFile=D:\elasticsearch\logs/hs_err_pid%p.log;-XX:MaxDirectMemorySize=2197815296;-XX:G1HeapRegionSize=4m ++JvmOptions -Delasticsearch;-Des.path.home="D:\elasticsearch\release";-Des.path.conf="D:\elasticsearch\config";-Des.distribution.flavor="default";-Des.distribution.type="zip";-Des.bundled_jdk="true" --LogPath "D:\elasticsearch\\release\..\logs\service" --LogPrefix "ElasticSearch" --StdError auto --StdOutput auto --PidFile "ElasticSearch.pid" --DisplayName "Elasticsearch 7.16.2 (ElasticSearch)" --Description "Elasticsearch 7.16.2 Windows Service - https://elastic.co" --Jvm "D:\elasticsearch\\release\jdk\bin\server\jvm.dll" --StartMode jvm --StopMode jvm --StartPath "D:\elasticsearch\release"  ++Environment HOSTNAME="%COMPUTERNAME%"`

After the fix :
`"D:\elasticsearch\release\bin\elasticsearch-service-x64.exe" //IS//ElasticSearch --Startup manual --StopTimeout 0 --StartClass org.elasticsearch.bootstrap.Elasticsearch --StartMethod main ++StartParams --quiet --StopClass org.elasticsearch.bootstrap.Elasticsearch --StopMethod close --Classpath "D:\elasticsearch\release\lib\*" --JvmMs 4192  --JvmMx 4192  --JvmSs 1024  --JvmOptions -Des.networkaddress.cache.ttl=60;-Des.networkaddress.cache.negative.ttl=10;-XX:+AlwaysPreTouch;-Djava.awt.headless=true;-Dfile.encoding=UTF-8;-Djna.nosys=true;-XX:-OmitStackTraceInFastThrow;-XX:+ShowCodeDetailsInExceptionMessages;-Dio.netty.noUnsafe=true;-Dio.netty.noKeySetOptimization=true;-Dio.netty.recycler.maxCapacityPerThread=0;-Dio.netty.allocator.numDirectArenas=0;-Dlog4j.shutdownHookEnabled=false;-Dlog4j2.disable.jmx=true;-Dlog4j2.formatMsgNoLookups=true;-Djava.locale.providers=SPI,COMPAT;--add-opens=java.base/java.io=ALL-UNNAMED;-XX:+UseG1GC;-XX:G1ReservePercent=25;-XX:InitiatingHeapOccupancyPercent=30;-Djava.io.tmpdir=D:\elasticsearch\\release\..\temp;-XX:+HeapDumpOnOutOfMemoryError;-XX:HeapDumpPath=data;-XX:ErrorFile=logs/hs_err_pid%p.log;-Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m;-Dlog4j2.formatMsgNoLookups=true;-XX:HeapDumpPath=D:\elasticsearch\data;-XX:ErrorFile=D:\elasticsearch\logs/hs_err_pid%p.log;-Xlog:gc*,gc+age=trace,safepoint:file=D:\elasticsearch\logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m;-XX:MaxDirectMemorySize=2197815296;-XX:G1HeapRegionSize=4m ++JvmOptions -Delasticsearch;-Des.path.home="D:\elasticsearch\release";-Des.path.conf="D:\elasticsearch\config";-Des.distribution.flavor="default";-Des.distribution.type="zip";-Des.bundled_jdk="true" --LogPath "D:\elasticsearch\\release\..\logs\service" --LogPrefix "ElasticSearch" --StdError auto --StdOutput auto --PidFile "ElasticSearch.pid" --DisplayName "Elasticsearch 7.16.2 (ElasticSearch)" --Description "Elasticsearch 7.16.2 Windows Service - https://elastic.co" --Jvm "D:\elasticsearch\\release\jdk\bin\server\jvm.dll" --StartMode jvm --StopMode jvm --StartPath "D:\elasticsearch\release"  ++Environment HOSTNAME="%COMPUTERNAME%"`

### CODE REVIEW

**Comment by rjernst** [comment]
@gillg It is unclear what problem you are trying to solve. Can you please describe the problem, and then lay out the solution you implement here?

**Comment by gillg** [comment]
The problem is described in the related issue.
Not 100% certain but I think any Java parameter which contains wildcard and/or coma, passed with *.options file or ES_JAVA_OPTS will be dropped in the installed service.
Out of the box use case, -XLog option in the jvm.options file.
You can easily test, if you launch elasticsearch.bat the option is used and appears in _nodes API for example.
But if you install ES as service with elasticsearch-service.bat the option is absent. To check deeper, take a look on "java" tab using `elasticsearch-service.bat manager elasticsearch`

File: distribution/src/bin/elasticsearch-service.bat [hunk 1]
```diff
@@ -177,6 +180,7 @@ for %%a in ("%ES_JAVA_OPTS:;=","%") do (
   )
   if "!other_opt!" == "true" set OTHER_JAVA_OPTS=!OTHER_JAVA_OPTS!;!var!
 )
+if defined loop_ES_JAVA_OPTS goto :loop
```
Comment (rjernst) [inline distribution/src/bin/elasticsearch-service.bat:183]: Can you explain (in a code comment) why this goto is necessary? Shouldn't the loop have already gone through all of the parameters?

**Review by rjernst — CHANGES_REQUESTED** [review]
Thanks for the PR! I've read through the linked issue. This PR, though, still needs a good description about what the change is doing, not just a "fixes" line. The bat code is already very cryptic, so additional comments explaining how this new loop around a for loop is working will help future developers be able to make changes here.

**Comment by rjernst** [comment]
@elasticmachine ok to test

**Comment by rjernst** [comment]
@elasticmachine update branch

**Comment by rjernst** [comment]
The elasticsearch-service batch file (along with other CLIs) were reimplemented in Java last year. I don't think this issue should exist anymore. If it still occurs for you in the latest release of Elasticsearch, please open a new issue and we will be happy to take a look.

**Pull request closed without merging**